src/test-jdk17/java/com/fasterxml/jackson/databind/records/JsonIdentityOnRecord5238Test.java

```diff
@@ -0,0 +1,83 @@
+package com.fasterxml.jackson.databind.records;
+
+import java.util.List;
+
+import org.junit.jupiter.api.Test;
+
+import com.fasterxml.jackson.annotation.*;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.testutil.DatabindTestUtil;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+// [databind#5238] immutable classes with @JsonIdentityInfo can be deserialized; records cannot
+public class JsonIdentityOnRecord5238Test
+        extends DatabindTestUtil
+{
+    // Record-based data
+    record ExampleRecord(List<ThingRecord> allThings, ThingRecord selected) { }
+
+    @JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator.class, property = "id")
+    record ThingRecord(int id, String name) { }
+
+    // POJO-based data
+    static class ExamplePojo {
+        public List<ThingPojo> allThings;
+        public ThingPojo selected;
+
+        @JsonCreator
+        public ExamplePojo(
+                @JsonProperty("allThings") List<ThingPojo> allThings,
+                @JsonProperty("selected") ThingPojo selected) {
+            this.allThings = allThings;
+            this.selected = selected;
+        }
+    }
+
+    @JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator.class, property = "id")
+    static class ThingPojo {
+        public final int id;
+        public final String name;
+
+        @JsonCreator
+        public ThingPojo(@JsonProperty("id") int id, @JsonProperty("name") String name) {
```

Comment:
I think this test should make sure `id` is only set via Constructor and NOT direct set on Field.
(JVM does allow forced settting of `final` Fields).

One way would be to make constructor change "id" value by prefix or suffix.

I know the end result does not look different, but it would explain discrepancy.